code/datums/components/decals/blood.dm

```diff
@@ -9,12 +9,15 @@
 
 /datum/component/decal/blood/generate_appearance(_icon, _icon_state, _dir, _layer, _color)
 	var/obj/item/I = parent
+	I.update_icon()
 	if(!_icon)
 		_icon = 'icons/effects/blood.dmi'
 	if(!_icon_state)
 		_icon_state = "itemblood"
 	var/icon = initial(I.icon)
-	var/icon_state = initial(I.icon_state)
+	var/icon_state = I.icon_state

```

Comment:
If you're going to make this support using current icon state then you need to update the decal whenever the icon changes. You'll want a signal in update_icon()